### ISSUE TITLE
Tasks and exceptions

### ISSUE THREAD

**Issue by Keno** [JuliaLang/julia#12485]
Working with exceptions in the presence of tasks can be a little frustrating. Two concrete issues I encountered:

```
julia> t = @schedule try
       while true; sleep(1); end
       catch e
       isa(e,InterruptException) || rethrow(e)
       end
Task (waiting) @0x00000001116c6cb0

julia> try
       error("You will never see me")
       catch e
       rethrow(e)
       finally
       Base.throwto(t,InterruptException)
       end
ERROR: InterruptException
 in throwto at ./task.jl:29
 in anonymous at ./no file:6
```

This is because exception in transit is not task local.

The other problem is that stack traces are pretty bad when `wait`ing on a task:

```
julia> foo() = error("Foo")
foo (generic function with 1 method)

julia> foo()
ERROR: Foo
 in foo at ./none:1

julia> wait(@schedule foo())
ERROR: Foo
 in yieldto at ./task.jl:24
 in wait at ./task.jl:313
 in wait at ./task.jl:228
 in wait at ./task.jl:61
```

The backtrace from the task is not shown anywhere. Ideally it would be displayed somehow.

**Comment by amitmurthy** [comment]
In the second case, "handled" tasks do not print their backtraces. However, it can still be accessed if you have a reference to the task object.

```
t=@schedule foo(); wait(t);
Base.show_backtrace(STDOUT, t.backtrace)
```

**Comment by amitmurthy** [comment]
In the first case, backtrace of `t` can be accessed by 
```
julia> Base.show_backtrace(STDOUT, t.backtrace)

 in yieldto at ./task.jl:75
 in wait at ./task.jl:371
 in wait at ./task.jl:286
 in sleep at stream.jl:616
 in anonymous at task.jl:2
```

The `Base.throwto` seems to hide the local rethrown exception. 
```
 try
              error("You will never see me")
              catch e
              rethrow(e)
              finally
              1
              end
ERROR: You will never see me
 in anonymous at no file:2
```

**Comment by amitmurthy** [comment]
Probably indirectly related - https://github.com/JuliaLang/julia/issues/1784

I have also seen a segfault triggered from code within a `finally` block  - trying to get a reduced case for the same.

**Comment by amitmurthy** [comment]
Also, `show` of a task in error now prints its backtrace.

```
julia> t=@schedule foo(); wait(t);
ERROR: Foo
 in yieldto at ./task.jl:75
 in wait at ./task.jl:371
 in wait at ./task.jl:286
 in wait at ./task.jl:112

julia> t
Task (failed) @0x00007f36d4160010
ErrorException("Foo")
 in foo at none:1
 in anonymous at task.jl:67
```

**Comment by vtjnash** [comment]
possible solution idea https://gist.github.com/vtjnash/9cd53434cc7f7e2f7a29

**Comment by JeffBezanson** [comment]
Yes, something like that should work. It can mostly be fixed in the front end by saving the exception data around the finally block. `jl_exception_in_transit` was intended to be very short-lived, and shouldn't have to be task local.

**Comment by JeffBezanson** [comment]
Also, need parens at the end of `Base.throwto(t,InterruptException())`.

**Comment by sbchisholm** [comment]
Is this similar?  I found that `rethrow()` (no arguments) will throw the last thrown exception even if it is not the exception that was caught in the catch statement:

```julia
type IgnoredException <: Exception end
type HandledException <: Exception end

try
    @async begin
        sleep(1.0)
        try 
            throw(IgnoredException())
        catch ex
            println("Task threw and caught $ex")
            !isa(ex, IgnoredException) && rethrow()
        end 
    end 

    println("Throwing HandledException")
    throw(HandledException())
catch ex
    if isa(ex, HandledException)
        println("Caught: $ex")
    end 
    sleep(2.0)
    println("Rethrowing...")
    rethrow(ex) # throws HandledException()
    # rethrow()  # throws IgnoredException()
end
```

If I don't pass `ex` to rethrow, it will throw the `IgnoredException()` that was caught and ignored by the asynchronous task.  I was expecting that `rethrow()` would throw the exception that was caught by the catch statement.

**Comment by JeffBezanson** [comment]
Yes that looks like the same issue to me.

**Comment by sbchisholm** [comment]
Thanks @JeffBezanson.

**Comment by c42f** [comment]
Experimental PR which fixes this: #28878
